### PR TITLE
New contract state

### LINKED ISSUE #29: New contract state: cargo-concordium
**Task description**

This builds on #28 

The main change that will affect cargo concordium is to do with local contract execution. Cargo-concordium needs to implement the same kind of backing radix tree that is used by the node so that it can execute contracts. Ideally, it would use the exact tree used by the node, however, this is not possible due to it being in Haskell. The remaining changes, if time permits, are going to be to add the ability to inspect the state and support for constructing custom states to run contracts in.

The remaining features are covered by #28 .

### LINKED ISSUE #30: New contract state: Execution engine support
**Task description**

Building on the V1 host developed in #27 the host functions dealing with state will have to be revised to use the new state implementation.

**Sub-tasks**
- [ ] Implement the new host
- [ ] Document the interface exposed to the smart contracts.
- [ ] Expose the process/init/receive functions via FFI and integrate into the node.

### LINKED ISSUE #39: Contract fallback methods: cargo-concordium support
**Task description**

Support invoking fallback methods in cargo-concordium. Concretely this means that `cargo concordium update` should check for the existence of a fallback and invoke it if it exists, even if an exact match is not found.

### PR DESCRIPTION
**Pull request by limemloh**
## Purpose

New state implementation for V1 contracts. This builds upon Concordium/concordium-wasm-smart-contracts#24 and
- Resolves Concordium/concordium-wasm-smart-contracts#29
- Resolves Concordium/concordium-wasm-smart-contracts#30
- Resolves Concordium/concordium-wasm-smart-contracts#39 

## Changes in wasm-chain-integration

- Provide a new implementation of contract state based on a radix tree with branching factor 16 (see Concordium/concordium-wasm-smart-contracts#38 for more details and rationale)
- Replace the state in V1 contracts with the new state implementation. This means changes in the validation of contracts to support new host functions, and implementation of handlers of host functions.
- For V1 contracts, keep track of whether the state has changed during execution. The definition of "state has changed" is whether any state modifying functions were called.
- Revised FFI for V1 contracts due to new handling of the state.

## Changes in cargo-concordium

- support for loading and storing the new state
- support for printing the new state, either at the end of execution, or as a standalone command

## Changes

Depends on:
- [x] https://github.com/Concordium/concordium-wasm-smart-contracts/pull/24
- [x] https://github.com/Concordium/concordium-contracts-common/pull/36

## Checklist

- [x] My code follows the style of this project.
- [x] The code compiles without warnings.
- [x] I have performed a self-review of the changes.
- [x] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [x] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 1]
```diff
@@ -0,0 +1,3053 @@
...
+
+impl<V> CachedRef<V> {
+    pub(crate) fn load_and_cache<F: BackingStoreLoad>(&mut self, loader: &mut F) -> &mut V
+    where
+        V: Loadable, {
+        match self {
+            CachedRef::Disk {
+                key,
+            } => {
+                let value = V::load_from_location(loader, *key).unwrap(); // TODO: Error handling.
```
Comment (EmilLa1) [inline wasm-chain-integration/src/v1/trie/low_level.rs:362]: Resolve now or later?

**Review by MilkywayPirate — COMMENTED** [review]
Looks good. 
A few questions and remarks of things we need to handle now or postpone.

File: cargo-concordium/src/main.rs [hunk 2]
```diff
@@ -752,10 +803,23 @@ fn handle_run_v1(run_cmd: RunCommand, module: &[u8]) -> anyhow::Result<()> {
...
     };
-    let print_result = |state: v0::State, logs: v0::Logs| -> anyhow::Result<()> {
-        print_logs(logs);
+    let print_state = |mut state: v1::trie::MutableState,
+                       loader: &mut v1::trie::Loader<&[u8]>,
+                       should_display_state: bool|
+     -> anyhow::Result<()> {
+        let mut collector = v1::trie::SizeCollector::default();
+        let frozen = state.freeze(loader, &mut collector);
+        // TODO: Decide on terminology to use for this so that it is clear what it
```
[the remark this answers is not in the document]
Comment (abizjak) [reply cargo-concordium/src/main.rs:812]: I revised the phrasing of the message.

File: wasm-chain-integration/src/v1/trie/api.rs [hunk 3]
```diff
@@ -0,0 +1,304 @@
...
+pub struct MutableStateInner {
+    /// Current root of the tree. The current generation.
+    root:      u32,
+    /// The mutable trie itself. The root is an index in the array of generation
+    /// roots.
+    /// The idea is that the mutex is acquired at the start of execution of the
+    /// contract and released at the end.
+    /// The reason for the mutex is that we need to be able to clone this so
+    /// that we can share it inside the single transaction.
+    pub state: Arc<Mutex<MutableTrie<Value>>>,
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/api.rs]: Ok.

File: wasm-chain-integration/src/v1/trie/low_level.rs (hunk 1 above)
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs]: Added precondition.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 4]
```diff
@@ -0,0 +1,3053 @@
...
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+/// A stem is a sequence of [Chunk]s. It is an optimization of the node
+/// representation where common parts of the key are stored inline in a single
+/// node instead of having many nodes with a single child.
+struct Stem {
+    // TODO: Split into short + long, or use smallvec/tinyvec, but make sure to benchmark any
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs]: Will make issue.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 5]
```diff
@@ -0,0 +1,3053 @@
...
+    pub(crate) data:         Box<[u8]>,
+    /// Whether the last chunk is partial or not.
+    pub(crate) last_partial: bool,
+}
+
+#[derive(Debug, Clone)]
+/// A mutable version of the stem.
+/// This is an implementation detail of the iterator.
+struct MutStem {
+    // TODO: Split into short + long, or use smallvec/tinyvec, but make sure to benchmark any
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs]: Will make issue.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 6]
```diff
@@ -0,0 +1,3053 @@
...
+        FollowStem::StemIsPrefix {
+            key_step,
+        }
+    } else {
+        FollowStem::Equal
+    }
+}
+
+impl<V: Clone> Node<V> {
+    /// TODO: This is not very efficient. It involves cloning nodes, which is
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs]: Removed todo.

File: wasm-chain-integration/src/v1/ffi.rs [hunk 7]
```diff
@@ -194,34 +218,61 @@ unsafe extern "C" fn call_receive_v1(
     output_len: *mut size_t,
 ) -> *mut u8 {
     let artifact = Arc::from_raw(artifact_ptr);
-    let res = std::panic::catch_unwind(|| {
-        let receive_ctx = v0::deserial_receive_context(slice_from_c_bytes!(
+    let res = std::panic::catch_unwind(|| -> *mut u8 {
+        // For FFI we only pass v0 contexts to keep the other end simpler.
+        let receive_ctx_common = v0::deserial_receive_context(slice_from_c_bytes!(
```
[the remark this answers is not in the document]
Comment (EmilLa1) [reply wasm-chain-integration/src/v1/ffi.rs:225]: Nevermind. I was just confused that v1 used the v0 namespace for deserializing. I do see now that the v0:ReceiveContext is common for both v0 and v1.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 8]
```diff
@@ -0,0 +1,3164 @@
...
+        have_removed
+    }
+}
+
+#[derive(Debug)]
+/// A link to a shared occurrence of a value V.
+/// This is used in this module to construct trees, allowing for sharing of
+/// values in trees and subtrees in case of the persistent tree.
+pub struct Link<V> {
+    link: Arc<RwLock<V>>,
```
Comment (Bargsteen) [inline wasm-chain-integration/src/v1/trie/low_level.rs:236]: Not sure I understand why the atomic part is needed and also the rwlock. Will revisit later, once I'm all the way through.

**Review by Bargsteen — CHANGES_REQUESTED** [review]
Still missing a deeper review of certain parts, but I'll post the comments now.

In general, I think the code looks good, but the big picture is hard to grasp.
I've been involved in some of the design decisions and overheard other comments about it, so I had some idea about how things were meant to work (fx the ideas behind freeze/thaw and generations), but I still found it hard to follow.

Some general documentation about the flow and design decisions would be greatly appreciated 😊 🙏  
- What happens when a contract entrypoint is invoked? What are the steps?
- What does a rollback look like and when does it happen?
- What is the idea behind the big concepts such as persistent/mutable state and generations.

Perhaps it could be explained from the standpoint of a typical flow with no optimizations and moving towards more optimizations, so you can see why the optimizations are there.

File: wasm-chain-integration/src/constants.rs [hunk 9]
```diff
@@ -40,15 +40,15 @@ pub const BASE_SIMPLE_TRANSFER_ACTION_COST: u64 = BASE_ACTION_COST + 40000;
 /// contract state) to the Wasm memory.
 #[inline(always)]
 pub fn copy_from_host_cost(x: u32) -> u64 { 10 + u64::from(x) }
-/// Cost of copying the given amount of bytes from the host (e.g., parameter or
+/// Cost of copying the given amount of bytes to the host (e.g., parameter or
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/constants.rs:44]: I have added some more documentation. I hope that reading this additional documentation gives enough insight into how the costs are determined, even if this is not explicit for each and every cost.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 10]
```diff
@@ -0,0 +1,3164 @@
...
+#[inline(always)]
+fn write_node_path_and_value_tag(
+    stem_len: usize,
+    no_value: bool,
+    out: &mut impl Write,
+) -> Result<(), std::io::Error> {
+    let value_mask: u8 = if no_value {
+        0
+    } else {
+        0b0100_0000
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:2768]: It is not so much that they are fixed relative to that, it is that "INLINE_STEM_LENGTH" has to fit into 6 bits.

These masks are just test the first two bits. I feel that introducing constants would only make it harder to read.

I will attempt to provide a comment where the tags are used.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 11]
```diff
@@ -0,0 +1,3164 @@
...
+            // Since we branch on 4 bits there can be at most 16 children.
+            // So using u8 is safe.
+            buf.write_u8(node.children.len() as u8)?;
+            for (k, _) in node.children.iter() {
+                buf.write_u8(k.value)?;
+                ref_stack.pop().unwrap().store(buf)?;
+            }
+            Ok(())
+        };
+        let mut ref_stack = Vec::<Reference>::new();
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:1427]: I have attempted to explain it.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 12]
```diff
@@ -0,0 +1,3164 @@
...
+            // store the value
+            if let Some(v) = &mut node.value {
+                let mut borrowed = v.borrow_mut();
+                match &mut *borrowed {
+                    InlineOrHashed::Inline(v) => {
+                        buf.write_u8(v[0])?;
+                        buf.write_all(&v[1..][0..usize::from(v[0])])?;
+                    }
+                    InlineOrHashed::Indirect(indirect) => {
+                        buf.write_u8(0b1111_1111u8)?;
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:1412]: It is now.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 13]
```diff
@@ -0,0 +1,3164 @@
...
+            let (stem_len, stem_ref) = node.path.to_slice();
+            write_node_path_and_value_tag(stem_len, node.value.is_none(), buf)?;
+            // store the path
+            buf.write_all(stem_ref)?;
+            // store the value
+            if let Some(v) = &mut node.value {
+                let mut borrowed = v.borrow_mut();
+                match &mut *borrowed {
+                    InlineOrHashed::Inline(v) => {
+                        buf.write_u8(v[0])?;
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:1408]: It is documented now. I do not want to add an explicit length field since it would increase the variant by 8 bytes which is not insignificant.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 14]
```diff
@@ -0,0 +1,3164 @@
...
+                ..
+            } => value.clone(),
+        }
+    }
+
+    /// Apply the supplied function to the contained value. The value is loaded
+    /// if it is not yet cached. Note that this will **not** cache the
+    /// value, the loaded value will be dropped.
+    #[inline(always)]
+    pub(crate) fn use_value<X>(
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:325]: Done.

File: wasm-chain-integration/src/v1/trie/api.rs [hunk 15]
```diff
@@ -0,0 +1,308 @@
...
+    pub fn serialize(
+        &self,
+        loader: &mut impl BackingStoreLoad,
+        out: &mut impl std::io::Write,
+    ) -> anyhow::Result<()> {
+        match self {
+            PersistentState::Empty => out.write_u8(0)?,
+            PersistentState::Root(ht) => {
+                out.write_u8(1)?;
+                ht.use_value_(loader, |loader, node| node.serialize(loader, out))?
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/api.rs:94]: I agree, I just did not know what the better name is. This pattern that I'm using here is not very rust like. It is very javascripty or Haskelly with supplying a callback to do stuff.

Instead of renaming, I have changed these `use_value` methods to `.get` methods that return something like `Cow` (but not quite that due to needless constraints of that type).

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 16]
```diff
@@ -0,0 +1,3164 @@
...
+        }
+    }
+}
+
+#[derive(Debug)]
+/// A persistent node. Cloning this is cheap, it only copies pointers and
+/// increments reference counts.
+pub struct Node {
+    /// Since a single node owns each value using Hashed<Cached<V>>
+    /// here makes sense, it makes it so that the hash is stored inline.
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:916]: Indeed it is. Obsolete comments mixing various previous versions of the structure :)

File: wasm-chain-integration/src/v1/trie/tests.rs [hunk 17]
```diff
@@ -0,0 +1,1027 @@
...
+        let (trie, mut loader) = make_mut_trie(inputs);
+        let mut frozen = if let Some(t) = trie.freeze(&mut loader, &mut EmptyCollector) {
+            t
+        } else {
+            ensure!(reference.is_empty(), "Reference map is empty, but trie is not.");
+            return Ok(());
+        };
+        let mut ser = Vec::new();
+        let _ = frozen.store_update(&mut ser);
+        ensure!(frozen.get(&mut loader).data.is_stored(), "Not all data is stored.");
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/tests.rs:161]: Indeed.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 18]
```diff
@@ -0,0 +1,3164 @@
...
+        value: V,
+    },
+    Cached {
+        key:   Reference,
+        value: V,
+    },
+}
+
+/// The default hash implementation is not a valid value.
+impl<V> Default for CachedRef<V> {
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:292]: It was used in the `deserialize` method via the `std::mem::take` method. I have changed that to use `mem::replace` instead with a more explicit dummy value, and removed the default.

File: wasm-chain-integration/src/v1/trie/api.rs [hunk 19]
```diff
@@ -0,0 +1,308 @@
...
+pub struct MutableStateInner {
+    /// Current root of the tree. The current generation.
+    root:  u32,
+    /// The mutable trie itself. The root is an index in the array of generation
+    /// roots.
+    /// The idea is that the mutex is acquired at the start of execution of the
+    /// contract and released at the end.
+    /// The reason for the mutex is that we need to be able to clone this so
+    /// that we can share it inside the single transaction.
+    state: Arc<Mutex<MutableTrie>>,
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/api.rs:185]: We need to be able to mutate the state, so we need the mutex. If we just had `Arc<MutableTrie>` then it is never possible to get a mutable reference to `MutableTrie`, and we need that.

`Arc<Mutex<T>>` is the equivalent to `Rc<RefCell<T>>`.

While it is true that execution is single-threaded the general execution context is multithreaded since (in general at least) we are using a parallel garbage collector. 

In this specific case I think it might be possible to use `Rc<RefCell<T>>` instead, but we'd need to carefully analyze that we don't have issues with cleanup (which is triggered via the Haskell GC in some cases). Since this does not impact performance I don't think the risks and time spent analyzing to make sure it would work are worth it.

We need multiple copies of this due to generations and rollback.

File: wasm-chain-integration/src/v1/trie/low_level.rs (hunk 8 above)
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:238]: I have added this comment to attempt to explain why Arc and RwLock are needed.
```
/// This [Link] achieves the following properties
/// - it is cheap to clone
/// - it allows for inner mutability
/// - it is safe to use in a concurrent context.
///
/// The cheap cloning is necessary to have efficient persistent trees. The
/// [Link] is used to point to children, as well as values and in the persistent
/// tree, modifying the value at a given point means we need to create copies of
/// the spine of the tree up to the root. This involves cloning pointers to any
/// parts that have not been modified. Thus these have to be cheap, so we use an
/// [Arc].
///
/// Inner mutability is needed so that the persistent tree may be loaded from
/// disk, as well as written to disk. This is achieved in combination with
/// [CachedRef]. Instead of using an [RwLock] we could instead use a [Mutex]
/// which would achieve the same API. However based on benchmarks the RwLock has
/// negligible overhead in the case of a single reader, and thus the [RwLock]
/// seems the more natural choice since it allows concurrent reads of the tree.
///
/// Finally, the reference counting must be atomic since parts of the tree are
/// dropped concurrently. While all the operations of the smart contract
/// execution engine are sequential, multiple states may be derived from the
/// same state in different threads. Currently this happens when using
/// invokeContract, but in the future with parallel block execution it might
/// also happen during normal block processing.
/// Additionally, if the Haskell runtime is configured with the parallel garbage
/// collector then parts of the tree might be dropped concurrently. This also
/// requires atomic reference counting.
```

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 20]
```diff
@@ -0,0 +1,3164 @@
...
+        V: AsRef<[u8]>, {
+        match self {
+            CachedRef::Disk {
+                key,
+            } => key.store(buf),
+            CachedRef::Memory {
+                value,
+            } => {
+                let key = backing_store.store_raw(value.as_ref())?;
+                let value = std::mem::replace(self, CachedRef::Disk {
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:434]: Ok.

This is not the most ergonomic part of rust.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 21]
```diff
@@ -0,0 +1,3164 @@
...
+
+/// Link to a value.
+type ValueLink = Link<InlineOrHashed>;
+
+#[derive(Debug)]
+/// A value that is either stored inline without a hash, or with a cached hash
+/// and a pointer to the actual value. The intention is that small values will
+/// be stored inline and larger values behind and indirection.
+pub enum InlineOrHashed {
+    Inline([u8; INLINE_VALUE_LEN + 1]),
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:823]: Ok.

The reason inlining into the byte array is to make sure no extra space is used. Due to alignment and cleverness of the compiler there seems to be no difference in the size of the type when factoring the value out. But it is more implicit.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 22]
```diff
@@ -0,0 +1,3164 @@
...
+        Ok(Node {
+            value,
+            path,
+            children: branches,
+        })
+    }
+}
+
+impl Node {
+    /// The entire tree in memory.
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:1334]: I chose to elaborate a bit more.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 23]
```diff
@@ -0,0 +1,3164 @@
...
+    pub fn store_update<S: BackingStoreStore>(
+        &mut self,
+        backing_store: &mut S,
+    ) -> Result<Vec<u8>, WriteError> {
+        let mut buf = Vec::new();
+        self.store_update_buf(backing_store, &mut buf)?;
+        Ok(buf)
+    }
+
+    pub fn store_update_buf<S: BackingStoreStore, W: std::io::Write>(
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:1410]: The problem with stack is that it is very implicit, and limited (by default I believe 8MB per thread on my machines at least, not sure about other platforms). This means it is very hard to argue for correctness in cases where the inputs (the tree) can be constructed maliciously. Another issue is that it matters what the call stack is before this function is called, so the available stack space for recursive calls might actually be much less.

Using an explicit stack (with vectors) is much more predicable, explicit, and safe, i.e., we can guarantee that we don't have to worry about that limit.

Regarding performance, I would not expect that using an explicit stack (i.e., without recursion in the language) would be faster.

I agree that it is harder to follow, but it needs to work. I also think that once you get a hang of rewriting recursive functions into functions without recursion and explicit control stack this is a fairly mechanical transformation.

[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:1410]: In any case, there is more documentation of the approach, as well as details in this function.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 24]
```diff
@@ -0,0 +1,3164 @@
...
+        } else {
+            let value_idx = values.len();
+            values.push(value);
+            *entry = Entry::Mutable {
+                entry_idx: value_idx,
+            }
+        };
+    }
+
+    pub fn next<L: BackingStoreLoad, C: TraversalCounter>(
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:1908]: Done.

File: wasm-chain-integration/src/v1/trie/api.rs [hunk 25]
```diff
@@ -0,0 +1,308 @@
...
+    }
+
+    /// Get a fresh mutable state generation. Modifications on this generation
+    /// will not be reflected in the current one. To resume execution of the
+    /// previous generation the inner state must first be normalized to the
+    /// previous generation.
+    pub fn make_fresh_generation(&mut self, loader: &mut impl BackingStoreLoad) -> Self {
+        if let Some(inner) = self.inner.as_mut() {
+            let mut trie = inner.lock();
+            trie.normalize(inner.root);
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/api.rs:258]: I am not sure I understand this question. Since the MutableTrie is shared, during rollback we do not actually clean it up.

File: wasm-chain-integration/src/v1/trie/types.rs [hunk 26]
```diff
@@ -0,0 +1,445 @@
...
+        Self {
+            inner: file,
+        }
+    }
+}
+
+impl<'a, A: AsRef<[u8]>> BackingStoreLoad for Loader<A> {
+    type R = Vec<u8>;
+
+    // FIXME: This is inefficient. We allocate too many vectors.
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/types.rs:216]: Resolved by using tinyvec

File: wasm-chain-integration/src/v1/trie/types.rs [hunk 27]
```diff
@@ -0,0 +1,445 @@
...
+#[derive(Default)]
+pub struct SizeCollector {
+    num_bytes: u64,
+}
+
+impl SizeCollector {
+    pub fn collect(self) -> u64 { self.num_bytes }
+}
+
+// TODO: Make sure this is adequate. There is a bit of overhead with size length
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/types.rs:129]: Now.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 28]
```diff
@@ -0,0 +1,3053 @@
...
+                }
+            }
+            Entry::Mutable {
+                entry_idx,
+            } => return values.get(entry_idx).map(f),
+            Entry::Deleted => None,
+        }
+    }
+
+    /// TODO: It might be useful to return a list of new nodes so that they
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:2188]: I elaborated a bit for future maintainers.

File: wasm-chain-integration/src/v1/trie/types.rs [hunk 29]
```diff
@@ -0,0 +1,445 @@
...
+    type Err = NoError;
+
+    #[inline(always)]
+    fn allocate(&mut self, _data: &V) -> Result<(), Self::Err> { Ok(()) }
+}
+
+/// A type that can be used to collect auxiliary information while a mutable
+/// trie is being frozen. Particular use-cases of this are collecting the size
+/// of new data, as well as new persistent nodes.
+/// TODO: Methods should probably return Result/Option so that we can terminate
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/types.rs:97]: I explained what could be done, and that what we do now is not too much of an issue.

If we have time in the future we can do it.

File: wasm-chain-integration/src/v1/trie/api.rs [hunk 30]
```diff
@@ -0,0 +1,308 @@
+//! The high-level API that is meant to be used by the rest of the project.
+//! It exposes two main constructs, the [PersistentState] and [MutableState]
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/api.rs:2]: I have added a few paragraphs.
```rust

//! The [`PersistentState`] is immutable. It is "persistent" in the sense of
//! persistent data structures, meaning that we never modify it in-place, but
//! always create new versions. This state is used for recording the state of a
//! smart contract at the end of a transaction execution.
//!
//! State modifications are done via the [`MutableState`] which exists during
//! transaction execution. This type is modified in-place and supports rollbacks
//! which are sometimes necessary during the transaction due to contract calls.
//!
//! The overall flow is that a [`PersistentState`] is
//! [thawed](PersistentState::thaw) into a [`MutableState`] when the contract
//! starts execution. Then a contract is executed, and if it terminates with
//! success the resulting state is [frozen](MutableState::freeze) into a
//! [`PersistentState`]. During this process only the part of the original state
//! that has been modified is copied. The new [`PersistentState`] in general
//! retains pointers to parts of the original state. This allows for efficient
//! state updates where we only have to store the parts of the state that are
//! new.
```

File: wasm-chain-integration/src/v1/trie/api.rs [hunk 31]
```diff
@@ -0,0 +1,308 @@
...
+
+#[derive(Debug, Clone)]
+/// This type is a technical device to support lazy conversion of
+/// [PersistentState] to [MutableState]. It contains the runtime representation
+/// of the tree that supports efficient updates.
+/// Clone on this structure is designed to be cheap, **and it is a shallow
+/// copy**. Modifications of the inner MutableTrie on either the clone or the
+/// original propagate to the other.
+pub struct MutableStateInner {
+    /// Current root of the tree. The current generation.
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/api.rs:180]: I have tried to explain it now.

File: wasm-chain-integration/src/v1/trie/types.rs [hunk 32]
```diff
@@ -0,0 +1,471 @@
...
+/// early in case of out of energy.
+pub trait Collector<V> {
+    fn add_value(&mut self, data: &V);
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/types.rs:100]: Done.

File: wasm-chain-integration/src/v1/trie/types.rs [hunk 33]
```diff
@@ -0,0 +1,471 @@
...
+/// Length of the stem that will be stored inline, i.e.,
+/// the stem length will be encoded in the tag bit, as opposed to separate
+/// 4 bytes.
+pub(crate) const INLINE_STEM_LENGTH: usize = 0b0011_1111;
+
+/// A trait that supports keeping track of resources during tree traversal, to
+/// make sure that resource bounds are not exceeded.
+pub trait TraversalCounter {
+    type Err: std::fmt::Debug;
+    fn tick(&mut self, num: u64) -> Result<(), Self::Err>;
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/types.rs:63]: Added (and also renamed not to confused with the energy `.tick`).

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 34]
```diff
@@ -0,0 +1,3164 @@
...
+            self.last_partial = false;
+        } else {
+            self.data.truncate(len / 2 + 1);
+            *self.data.last_mut().expect("Odd implies at least one element.") &= 0xf0;
+            self.last_partial = true;
+        }
+    }
+
+    /// Extend the current stem by the content of the given stem.
+    fn extend(&mut self, second: &Stem) {
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:577]: Added.

File: wasm-chain-integration/src/v1/mod.rs [hunk 35]
```diff
@@ -315,12 +353,280 @@ mod host {
...
+        stack: &mut machine::RuntimeStack,
+        energy: &mut InterpreterEnergy,
+        state: &mut InstanceState<BackingStore>,
+    ) -> machine::RunResult<()> {
+        let key_len = unsafe { stack.pop_u32() };
+        let key_start = unsafe { stack.pop_u32() } as usize;
+        let key_end = key_start + key_len as usize;
+        // this cannot overflow on 64-bit platforms, so it is safe to just add
+        ensure!(key_end <= memory.len(), "Illegal memory access.");
+        let key = &memory[key_start..key_end];
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/mod.rs:431]: I prefer not to do that. All of these methods are structured in a similar way with arguments being parsed inline.

I do not feel that factoring out this part would bring much value.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 36]
```diff
@@ -0,0 +1,3164 @@
...
+            entries:         Vec::new(),
+        }
+    }
+
+    /// Check whether the current generation is an empty tree.
+    pub fn is_empty(&self) -> bool { self.generations.last().map_or(false, |x| x.root.is_none()) }
+}
+
+impl MutableTrie {
+    pub fn new_generation(&mut self) {
```
[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:1862]: > Also on how it relates to `normalize`. Should `normalize` always be called prior to `new_generation`?

No. Normalize just forgets 0 or more of the most recent generations. I have added some documentation that hopefully explains this.

[the remark this answers is not in the document]
Comment (abizjak) [reply wasm-chain-integration/src/v1/trie/low_level.rs:1912]: I have added more documentation both to the api module and the low_level module outlining the overall approach.

**Comment by abizjak** [comment]
@Bargsteen I have added a bunch more documentation now. I hope some of the high-level concepts are clearer now.

There could be more detailed documentation, but I need to focus on merging this and the node changes in. So let me know if there is anything that is critically missing.

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 37]
```diff
@@ -0,0 +1,3391 @@
...
+/// This is used in this module to construct trees, allowing for sharing of
+/// values in trees and subtrees in case of the persistent tree.
+///
+/// This [Link] achieves the following properties
+/// - it is cheap to clone
+/// - it allows for inner mutability
+/// - it is safe to use in a concurrent context.
+///
+/// The cheap cloning is necessary to have efficient persistent trees. The
+/// [Link] is used to point to children, as well as values and in the persistent
```
Comment (Bargsteen) [inline wasm-chain-integration/src/v1/trie/low_level.rs:275]:
```suggestion
/// [Link] is used to point to children, as well as values, in the persistent
```

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 38]
```diff
@@ -0,0 +1,3391 @@
...
+/// tree, modifying the value at a given point means we need to create copies of
```
Comment (Bargsteen) [inline wasm-chain-integration/src/v1/trie/low_level.rs:276]:
```suggestion
/// tree. Modifying the value at a given point means we need to create copies of
```

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 39]
```diff
@@ -0,0 +1,3391 @@
...
+///
+/// Finally, the reference counting must be atomic since parts of the tree are
+/// dropped concurrently. While all the operations of the smart contract
+/// execution engine are sequential, multiple states may be derived from the
+/// same state in different threads. Currently this happens when using
+/// invokeContract, but in the future with parallel block execution it might
+/// also happen during normal block processing.
+/// Additionally, if the Haskell runtime is configured with the parallel garbage
+/// collector then parts of the tree might be dropped concurrently. This also
+/// requires atomic reference counting.
```
Comment (Bargsteen) [inline wasm-chain-integration/src/v1/trie/low_level.rs:297]: 👌

File: wasm-chain-integration/src/v1/trie/low_level.rs (hunk 8 above)
Comment (Bargsteen) [reply wasm-chain-integration/src/v1/trie/low_level.rs:299]: 👌

File: wasm-chain-integration/src/v1/trie/low_level.rs [hunk 40]
```diff
@@ -1382,15 +1515,34 @@ impl Node {
...
     pub fn store_update_buf<S: BackingStoreStore, W: std::io::Write>(
         &mut self,
         backing_store: &mut S,
         buf: &mut W,
     ) -> StoreResult<()> {
+        // This function would be very natural to write recursively, essentially
+        // - recursively write all the children that are only in memory
+        // - this would return Reference's to where these children are written in the
+        //   backing store
+        // - use these to write the node into the provided buffer (`buf`).
```
Comment (Bargsteen) [inline wasm-chain-integration/src/v1/trie/low_level.rs:1531]: :ok_hand:

File: wasm-chain-integration/src/v1/trie/types.rs [hunk 41]
```diff
@@ -53,14 +53,17 @@ pub type LoadResult<A> = Result<A, LoadError>;
 
 /// Length of the stem that will be stored inline, i.e.,
 /// the stem length will be encoded in the tag bit, as opposed to separate
-/// 4 bytes.
+/// 4 bytes. This must always fit 6 bits, i.e., can be no more than 127.
```
Comment (Bargsteen) [inline wasm-chain-integration/src/v1/trie/types.rs:56]: 63?
```suggestion
/// 4 bytes. This must always fit 6 bits, i.e., can be no more than 63.
```

**Review by Bargsteen — APPROVED** [review]
The added documentation was much appreciated 🎉 
LGTM 👍
